### ISSUE TITLE
Influence on the order of several APIC tags

### ISSUE THREAD

**Issue by charlydelta** [quodlibet/mutagen#339]
I use mutagen to make a linting script for my MP3 music library, which has these requirements:

1. An embedded picture of the album cover
1. An embedded picture of the artist/band logo
1. An embedded picture of the artist/band
1. That the order of these pictures is as in this list, so the first APIC in the file should be the album cover, then the logo, then the artist. The order of the latter two pictures can be swapped, but it is important that the first picture is the album cover since most music players do not honour the album cover type "3". Instead, they will just take the first APIC tag and assume that's the album cover.
1. No other embedded pictures are allowed

I now have some issues:

1. I can't lint if the order of the pictures is correct, because I would have to use the `getall('APIC')` method. This method does *not* return consistent results - The order of the pictures in the list is random for each load (that is, every time I load the very same MP3 I have different results in the order of the list returned by `getall`).
1. I have absolutely no influence on the order in which the APIC tags are written - Therefore I am not able to solve potential issues in the MP3 file -, mutagen will **always** write the band logo first, then the artist picture, and the album cover as last.

Long story short - I am neither able to check whether the MP3 files are in the correct order, nor I am able to write the album covers in a specific order even if I know that the order is not correct.

I am not sure what the specification says about this. But I am able to do what I want using `eyeD3`, which honours the order in which the pictures are added with the `add` method and where the result of an APIC query is consistent with the order in which the tags occur in the file.

**Comment by lazka** [comment]
```
There is no fixed order of the frames' appearance in the tag,
although it is desired that the frames are arranged in order of
significance concerning the recognition of the file. An example of
such order: UFID, TIT2, MCDI, TRCK ...
```

We have some basic ordering for text frames right now and write APIC last because some sofotware as problems with large frames, but we could always do something smarter about APIC internal ordering.

How would you order them? https://github.com/quodlibet/mutagen/blob/master/mutagen/id3/_specs.py#L21

> most music players do not honour the album cover type "3"

Which for example?

**Comment by charlydelta** [comment]
@lazka thanks for your response

Let me elaborate: I use Debian 9 which ships with Nautilus as file manager, and rhythmox as music player. I also have VLC installed. All of these softwares display the *first* APIC they find - they don't bother searching for type=3. Also MP3Tag for windows shows the order in which they where saved by mutagen. 

My suggestion would be to save the embedded pictures either in the order in which they get added to the API (that's what eyeD3 does - or to save them sorted by type number ascending (the standard seems to have ordered the types by relevance. For the second approach, however, I would recommend an exception: always write type=3 first to the file in order to support legacy software, because you'd usually expect to have the album art displayed.

What do you guys think? Maybe the first approach would be the most natural approach in terms of API handling/usage.

**Comment by lazka** [comment]
I don't want to guarantee any order, but sorting by number and preferring 3 sounds good. The would help everyone and not just the users calling add() in the right order.

**Comment by lazka** [comment]
We have to make sure any APIC is still last because of #278

**Comment by charlydelta** [comment]
> The would help everyone and not just the users calling add() in the right order.

I trust you in your opinion, because I am certainly no expert in the ID3 standard - I just hope that this is okay, because without the reasoning it really sounds like a super arbitrary rule. 

Another thought regarding this: Have you considered throwing a lambda function into the `sorted()` function, which imitates the current behaviour so that backwards compatibility is guaranteed, and supply an API for the user to set/override this function? This would be the least invasive approach at least. This suggestion is of course not mutually exclusive with the default of sorting the APICs by type preferring 3.

Regarding the other issue: Why is the order of the APIC tags random when reading a file? Is it a design decision because it works faster that way, or could it be changed to reflect the order in which the APIC tags are listed in the MP3?

What I mean is this:

```python
song = MP3(...)
embedded_images = song.tags.getall('APIC')
```

For me, the order of `embedded_images` is completely random - I would expect it to reflect the order of the frames in the MP3. Otherwise I am not able to lint this requirement.

**Comment by lazka** [comment]
> I trust you in your opinion, because I am certainly no expert in the ID3 standard - I just hope that this is okay, because without the reasoning it really sounds like a super arbitrary rule.

You suggested it ;) But I don't see anything wrong with it.

> Another thought regarding this: Have you considered throwing a lambda function into the sorted() function, which imitates the current behaviour so that backwards compatibility is guaranteed, and supply an API for the user to set/override this lambda? This would be the least invasive approach at least. This suggestion is of course not mutually exclusive with the default of sorting the APICs by type preferring 3.

From a maintainer perspective adding API is very invasive, as that means I have to expose internals and keep it working forever.

> Regarding the other issue: Why is the order of the APIC tags random when reading a file? Is it a design decision because it works faster that way, or could it be changed to reflect the order in which the APIC tags are listed in the MP3?

Frames are stored/exposed through a dict interface, the getall() is just a helper which iterates over the keys. I assume it was chosen back then because each frame type has a unique key and frames have no order according to the standard, so a dict interface seemed like a good match.

> For me, the order of embedded_images is completely random - I would expect it to reflect the order of the frames in the MP3. Otherwise I am not able to lint this requirement.

If you use Python 3.6+ there is a good chance that getall() returns the right order, as dict preserves insertion order there and that should leak to getall(), but not 100% sure.

You could try reporting bugs with VLC/Rhythmox to prefer cover images. That would help all users :)

**Comment by charlydelta** [comment]
> You suggested it ;) But I don't see anything wrong with it.

I know, I just wanted to emphasize that I am no expert and cannot evaluate if the idea is good.

> From a maintainer perspective adding API is very invasive [...]

Oh, yeah I see your point. All right :)

> Frames are stored/exposed through a dict interface [...] If you use Python 3.6+ there is a good chance that getall() returns the right order [...]

Thank you very much for the information and the hints! I think I can work with this now :)

> You could try reporting bugs with VLC/Rhythmox to prefer cover images. That would help all users :)

I will do that - but most likely it's just an old version if ID3 they're using or these issues are already sorted out, considering that I am using Debian stretch (stable as of 2018).
